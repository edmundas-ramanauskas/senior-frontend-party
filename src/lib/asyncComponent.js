import React from 'react';
import PropTypes from 'prop-types';

class AsyncComponent extends React.Component {
  static propTypes = {
    load: PropTypes.func.isRequired,
  };

  state = { component: null };

  componentDidMount() {
    this.props.load().then(({ default: component }) => {
      this.setState({ component });
    });
  }

  render() {
    const { component: Component } = this.state;
    const { load, ...props } = this.props;
    return Component ? <Component {...props} /> : null;
  }
}

export default importComponent => props => (
  <AsyncComponent load={importComponent} {...props} />
);
