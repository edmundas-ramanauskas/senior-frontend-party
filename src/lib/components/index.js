export { default as Input } from './Input';
export { default as Wrapper } from './Wrapper';
export { default as Redirect } from './Redirect';
