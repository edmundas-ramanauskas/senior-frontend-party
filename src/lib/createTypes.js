import { pipe, reduce } from 'ramda';

export default (base, types, nameSpace = 'testio') =>
  pipe(
    reduce(
      (result, type) => ({ ...result, [type]: `${nameSpace}/${base}/${type}` }),
      {},
    ),
    Object.freeze,
  )(types);
