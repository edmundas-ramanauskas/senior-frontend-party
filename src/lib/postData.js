export default function postData(url, data) {
  return fetch(url, {
    body: JSON.stringify(data),
    headers: {
      'Content-Type': 'application/json',
    },
    method: 'POST',
  }).then(async response => {
    const body = await response.json();
    if (response.status !== 200) {
      return { error: body.message };
    }
    return body;
  });
}
