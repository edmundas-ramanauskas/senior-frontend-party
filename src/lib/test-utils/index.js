export { default as withReduxForm } from './withReduxForm';
export { default as withReduxStore } from './withReduxStore';
export { default as withStyletron } from './withStyletron';
