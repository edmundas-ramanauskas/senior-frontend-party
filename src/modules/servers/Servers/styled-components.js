import { styled } from 'styletron-react';
import { Button } from 'react-bootstrap';

const logo = require('./images/logo.png');
const logout = require('./images/logout.svg');
const logoutHover = require('./images/logout-hover.svg');

export const Logo = styled('span', {
  display: 'inline-block',
  width: '115px',
  height: '30px',
  background: `url(${logo}) no-repeat center center`,
});

export const Logout = styled(Button, {
  display: 'inline-block',
  paddingLeft: '28px',
  background: `url(${logout}) no-repeat left center`,
  backgroundSize: '16px',
  color: '#2f3254',
  ':hover': {
    color: '#99cc33',
    textDecoration: 'none',
    backgroundImage: `url(${logoutHover})`,
  },
  ':active': {
    outline: 'none !important',
  },
});

export const Row = styled('div', {
  display: 'flex',
  justifyContent: 'space-between',
  alignItems: 'center',
});

export const Head = styled(Row, {
  backgroundColor: '#f5f5f5',
  textTransform: 'uppercase',
});

export const Title = styled(Row, {
  padding: '26px 0',
});

export const Col = styled('div', ({ $fontSize = 16 }) => ({
  fontSize: `${$fontSize}px`,
  padding: '12px 15px',
}));
