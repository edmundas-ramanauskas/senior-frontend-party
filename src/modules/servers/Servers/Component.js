import React from 'react';
import PropTypes from 'prop-types';
import { Wrapper } from '../../../lib/components';
import { Col, Head, Logo, Logout, Title } from './styled-components';
import Item from './Item';

const Servers = ({ servers, onClickLogout }) => (
  <Wrapper>
    <Title>
      <Col>
        <Logo />
      </Col>
      <Col>
        <Logout bsStyle="link" onClick={onClickLogout}>
          Logout
        </Logout>
      </Col>
    </Title>
    <Head>
      <Col $fontSize={14}>Server</Col>
      <Col $fontSize={14}>Distance</Col>
    </Head>
    {servers.map(({ name, distance }) => (
      <Item key={`${name} - ${distance}`} {...{ name, distance }} />
    ))}
  </Wrapper>
);

Servers.propTypes = {
  servers: PropTypes.arrayOf(
    PropTypes.shape({
      name: PropTypes.string,
      distance: PropTypes.number,
    }),
  ).isRequired,
  onClickLogout: PropTypes.func.isRequired,
};

export default Servers;
