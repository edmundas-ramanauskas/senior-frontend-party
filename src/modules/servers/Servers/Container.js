import { ascend, isNil, pipe, prop, sortWith } from 'ramda';
import { connect } from 'react-redux';
import { branch, compose, mapProps, withHandlers } from 'recompose';
import { purgeToken, withToken } from '../../login';
import Servers from './Component';
import redirectToLogin from './redirectToLogin';
import withServers from '../withServers';

const getToken = prop('token');

const sortByDistanceAndName = sortWith([
  ascend(prop('distance')),
  ascend(prop('name')),
]);

export default compose(
  // get token from store
  withToken,
  branch(
    // if token is null
    pipe(getToken, isNil),
    // redirect to login
    redirectToLogin,
    // else
    compose(
      connect(null, {
        onLogout: purgeToken,
      }),
      withHandlers({
        onClickLogout: ({ onLogout }) => () => onLogout(),
      }),
      // fetch servers list
      withServers(getToken),
      mapProps(({ servers, ...props }) => ({
        servers: sortByDistanceAndName(servers),
        ...props,
      })),
    ),
  ),
)(Servers);
