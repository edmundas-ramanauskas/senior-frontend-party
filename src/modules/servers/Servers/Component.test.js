import React from 'react';
import renderer from 'react-test-renderer';
import { configure, shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import { withStyletron } from '../../../lib/test-utils';
import Servers from './Component';

configure({ adapter: new Adapter() });

const context = { styletron: { injectRawDeclaration: () => {} } };

describe('Servers', () => {
  it('Component renders corectly', () => {
    const ServersWithProviders = withStyletron(Servers);
    const component = renderer.create(
      <ServersWithProviders servers={[]} onClickLogout={() => {}} />,
    );
    expect(component.toJSON()).toMatchSnapshot();
  });

  it('All elements are rendered', () => {
    const props = {
      servers: [
        { name: 'Server1', distance: 100 },
        { name: 'Server2', distance: 100 },
      ],
      onClickLogout: () => {},
    };
    const servers = shallow(<Servers {...props} />, {
      context,
    });
    const items = servers.find('Item');
    expect(items).toHaveLength(2);
  });

  it('Logout callback is triggered', done => {
    shallow(<Servers servers={[]} onClickLogout={done} />, {
      context,
    })
      .find({ bsStyle: 'link' })
      .simulate('click');
  });
});
