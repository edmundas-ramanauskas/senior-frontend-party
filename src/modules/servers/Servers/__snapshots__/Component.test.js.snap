// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Servers Component renders corectly 1`] = `
<div
  className="container-fluid"
>
  <div
    className="row"
  >
    <div
      className="col-md-10 col-md-offset-1 col-xs-12"
    >
      <div
        className="undefined undefined undefined undefined"
      >
        <div
          className="undefined undefined"
        >
          <span
            className="undefined undefined undefined undefined"
          />
        </div>
        <div
          className="undefined undefined"
        >
          <button
            className="undefined undefined undefined undefined undefined undefined undefined undefined undefined btn btn-link"
            disabled={false}
            onClick={[Function]}
            type="button"
          >
            Logout
          </button>
        </div>
      </div>
      <div
        className="undefined undefined undefined undefined undefined"
      >
        <div
          className="undefined undefined"
        >
          Server
        </div>
        <div
          className="undefined undefined"
        >
          Distance
        </div>
      </div>
    </div>
  </div>
</div>
`;
